specs/agents/metrics.md

```diff
@@ -50,11 +50,109 @@ Agents should not send the `system.process.cgroup.memory.mem.limit.bytes` metric
+
+### Agent Health Metrics
+
+Agents SHOULD expose metrics which allow the analysis of the internal state of the agent, especially including the [reporting/transport mechanism](transport.md).
+
+Many of these metrics are `COUNTER`s: a `COUNTER` is a monotonically increasing metric.
+It is periodically reported by sending the difference of the counter-value now vs its value at the time of the previous report.
+In order to save elasticsearch disk space, zero-increases are not sent.
+
+Example:
+| **Time** | **Counter** | **Reported Metric Value** |
+|----------|-------------|---------------------------|
+| 0s       | 7           | 7                         |
+| 30s      | 10          | 3                         |
+| 1m       | 10          | 0 (not sent)              |
+| 1m30s    | 12          | 2                         |
+
+
+#### Event Count Metrics
+
+Agents send their data to the Intake API of the APM-server in the form of a stream of events (e.g. transactions, spans, metricsets...).
+
+The following metrics counting these events SHOULD be exposed:
+ - `agent.events.total`: `COUNTER` of events either submitted to the internal reporting queue or dropped upon submit attempt
+ - `agent.events.dropped.queue`: `COUNTER` of events where the attempt to add it to the reporting queue failed (e.g. due to a full queue)
+ - `agent.events.dropped.error`: `COUNTER` of events dropped after they previously have been successfully added to the queue.
+
+If an agent supports these metrics, the metrics MUST be captured for all events which would reach the APM-Server in an ideal, failure-free world.
+Therefore an event MUST be taken into account for these metrics if an attempt is made to submit it to the internal reporting queue, regardless of the success.
+Events dropped before this attempt (e.g. due to [sampling](tracing-sampling.md) or [transaction_max_spans](handling-huge-traces/tracing-spans-limit.md)) are NOT counted.
+
+The value of `agent.events.dropped.error` is an upper bound for the actual number of dropped events after entering the queue.
+For example, when an request to the Intake API fails, all events within this request MUST be considered as failed.
+In reality, it is possible that e.g. half of the data was actually successfully received and forwarded by the APM server.
+
+All these metrics MUST have the label `eventType`. The possible values for `eventType` are the lowercase names of the events according to the [spec](https://github.com/elastic/apm-server/tree/main/docs/spec/v2) (e.g. `transaction`, `metricset`) or the value `other`.
+
+Agents SHOULD attempt to assign the appropriate label value based on the counted event. If this would impose significant implementation overhead, the value `other` MUST be used instead.
+
+All event count metrics are enabled/disabled via the `agent_event_metrics` configuration:
```

Comment:
Do we need this if the metrics can also be disabled via `disable_metrics=agent.events.queue.*`?

Reply:
Ahh I was looking for the `disable_metrics` option but somehow failed to find it :D

In general why I added these configuration options per group of metrics was because some of them are opt-out while others are opt in. E.g. the queue-metrics are opt-out, the overhead-metrics are opt-in.

This means that for the overhead metrics we need something in addition to `disable_metrics` to enable them. For this reason I added the `agent_overhead_metrics` option. For the sake of consistency I also added the `agent_event_metrics` for the other metrics as well.

So I would enable the collection of an individual metric iff it is not matched by `disable_metrics` AND it's config-flag is `true` (e.g. `agent_overhead_metrics`). I will adjust this in the spec accordingly.

The remaining open question is whether we should remove the config flag `agent_event_metrics` for the other metrics? I'm free to do so if you don't see any consistency problem here.

Reply:
After a discussion in the agent weekly, we decided that `agent_event_metrics` is not required. `agent_overhead_metrics` is required for an opt-in.